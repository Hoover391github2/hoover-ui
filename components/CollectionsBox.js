import { Component } from 'react';
import PropTypes from 'prop-types';
import Loading from './Loading';

class Checkbox extends Component {
    render() {
        const { name, checked, title, onChange } = this.props;
        return (
            <div className="checkbox">
                <label>
                    <input
                        type="checkbox"
                        id={`checkbox-${name}`}
                        checked={checked}
                        onChange={e => onChange(name, e.target.checked)}
                    />{' '}
                    {title}
                </label>
            </div>
        );
    }
}

export default class CollectionsBox extends Component {
    static propTypes = {
        collections: PropTypes.arrayOf(
            PropTypes.shape({
                name: PropTypes.string.isRequired,
                title: PropTypes.string.isRequired,
            })
        ),
        selected: PropTypes.arrayOf(PropTypes.string).isRequired,
        counts: PropTypes.object,
    };

    changeSelection(selected) {
        this.props.onChange(selected);
    }

    handleChange(name, checked) {
        var all = this.props.collections.map(c => c.name);
        var selected = this.props.selected.splice(0);

        if (checked) {
            selected = [].concat(selected, [name]);
        } else {
            selected = selected.filter(c => c != name);
        }

        this.changeSelection(selected);
    }

    renderCheckboxes() {
        const { collections, selected, counts } = this.props;

        let allCheckbox = null;

        if (collections.length > 1) {
            let allSelected = true;
            for (let col of collections) {
                if (selected.indexOf(col.name) < 0) {
                    allSelected = false;
                }
            }

            let selectAll = () => {
                let selected = [];
                if (allSelected) {
                    this.changeSelection([]);
                } else {
                    this.changeSelection(collections.map(col => col.name));
                }
            };

            allCheckbox = (
                <Checkbox
                    name={'_all_'}
                    title={<em>All</em>}
                    key={'_all_'}
                    checked={allSelected}
                    onChange={selectAll}
                />
            );
        }

        const collectionCheckboxes = collections.map(col => (
            <Checkbox
                name={col.name}
                title={col.title + (counts ? ` (${counts[col.name] || 0})` : '')}
                key={col.name}
                checked={selected.indexOf(col.name) > -1}
                onChange={this.handleChange.bind(this)}
            />
        ));

        return (
            <div>
                {allCheckbox}
                {collectionCheckboxes}
            </div>
        );
    }

    render() {
        var result = null;
        let { selected, collections } = this.props;

        if (collections) {
            if (collections.length) {
                result = this.renderCheckboxes();
            } else {
                result = <em>no collections available</em>;
            }
        } else {
            result = <Loading />;
        }

        return (
            <div className="collections-box">
                <small className="text-muted">Collections</small>
                <div>{result}</div>
            </div>
        );
    }
}
