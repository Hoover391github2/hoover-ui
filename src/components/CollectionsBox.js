import Checkbox from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormGroup from '@material-ui/core/FormGroup';
import ListItem from '@material-ui/core/ListItem';
import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import PropTypes from 'prop-types';
import { Component } from 'react';
import { connect } from 'react-redux';
import { setCollectionsSelection } from '../actions';
import Loading from './Loading';
import { formatThousands } from '../utils';

const styles = theme => ({
    root: {
        display: 'flex',
    },
    formControlLabel: {
        width: '100%',
    },
    progress: {
        color: '#999',
        'font-size': '7.5pt',
    },
});

export class CollectionsBox extends Component {
    static propTypes = {
        classes: PropTypes.object.isRequired,
        collections: PropTypes.arrayOf(
            PropTypes.shape({
                name: PropTypes.string.isRequired,
                title: PropTypes.string.isRequired,
            })
        ),
        selected: PropTypes.arrayOf(PropTypes.string).isRequired,
        counts: PropTypes.object,
    };

    changeSelection(selected) {
        this.props.dispatch(setCollectionsSelection(selected));
    }

    handleChange = event => {
        const name = event.target.name;
        const checked = event.target.checked;

        var selected = this.props.selected.splice(0);

        if (checked) {
            selected = [].concat(selected, [name]);
        } else {
            selected = selected.filter(c => c != name);
        }

        this.changeSelection(selected);
    };

    renderCheckboxes() {
        const { collections, selected, counts, classes } = this.props;

        let allCheckbox = null;

        if (collections.length > 1) {
            const allSelected = collections.every(c => selected.includes(c.name));

            allCheckbox = (
                <FormControlLabel
                    className={classes.formControlLabel}
                    control={
                        <Checkbox
                            key={'_all_'}
                            checked={allSelected}
                            onChange={() =>
                                allSelected
                                    ? this.changeSelection([])
                                    : this.changeSelection(
                                          collections.map(c => c.name)
                                      )
                            }
                            value="_all_"
                        />
                    }
                    label="All"
                />
            );
        }

        const collectionCheckboxes = collections.map(col => (
            <FormControlLabel
                className={classes.formControlLabel}
                key={col.name}
                control={
                    <Checkbox
                        key={col.name}
                        name={col.name}
                        checked={selected.includes(col.name)}
                        onChange={this.handleChange}
                        value="_all_"
                    />
                }
                label={
                    <>
                        <span>
                            {col.title + (counts && counts[col.name] ? ` (${formatThousands(counts[col.name])})` : '')}
                        </span>
                        <i className={classes.progress}><br/>{col.stats.progress_str}</i>
                    </>
                }
            />
        ));

        return (
            <div>
                {allCheckbox}
                {collectionCheckboxes}
            </div>
        );
    }

    render() {
        let result;

        const {
            selected,
            collections,
            classes,
            isFetching,
            wasFetched,
        } = this.props;

        if (collections) {
            if (collections.length) {
                result = this.renderCheckboxes();
            } else {
                result =
                    isFetching || !wasFetched ? null : (
                        <Typography>no collections available</Typography>
                    );
            }
        } else {
            result = <Loading />;
        }

        return (
            <ListItem dense>
                <FormGroup>{result}</FormGroup>
            </ListItem>
        );
    }
}

const mapStateToProps = ({
    collections: { isFetching, items, selected, counts, wasFetched },
}) => ({
    isFetching,
    collections: items,
    selected,
    counts,
});

export default connect(mapStateToProps)(withStyles(styles)(CollectionsBox));
