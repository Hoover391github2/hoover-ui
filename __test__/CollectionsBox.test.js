import React from 'react';
import renderer from 'react-test-renderer';
import CollectionsBox from '../components/CollectionsBox';

it('renders the collections box with a collection', () => {
    const tree = renderer
        .create(
            <CollectionsBox
                selected={['testcollection']}
                collections={[{ name: 'testcollection', title: 'Test Collection' }]}
            />
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});

it('renders a loading collections box', () => {
    const tree = renderer.create(<CollectionsBox selected={[]} />).toJSON();

    expect(tree).toMatchSnapshot();
});
