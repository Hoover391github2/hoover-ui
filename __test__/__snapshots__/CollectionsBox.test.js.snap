// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a loading collections box 1`] = `
<div
  className="collections-box"
>
  <small
    className="text-muted"
  >
    Collections
  </small>
  <div>
    <div
      className="loading"
    >
      <i
        aria-hidden="true"
        className="fa fa-spinner loading-animate"
      />
    </div>
  </div>
</div>
`;

exports[`renders the collections box with a collection 1`] = `
<div
  className="collections-box"
>
  <small
    className="text-muted"
  >
    Collections
  </small>
  <div>
    <div>
      <div
        className="checkbox"
      >
        <label>
          <input
            checked={true}
            id="checkbox-testcollection"
            onChange={[Function]}
            type="checkbox"
          />
           
          Test Collection
        </label>
      </div>
    </div>
  </div>
</div>
`;
